### PR TITLE
Add user auth

### LINKED ISSUE #32: [Backend]: User login and logout
### Description

As a developer I want a endpoint to log in and log out users.

### Acceptance Criteria

A endpoint to log in and log out a user. Potentially with separate access and refresh tokens and the respective endpoints to handle them.

### Comment

_No response_

### LINKED ISSUE #89: [Bug Report]: User Migration Passwords are not hashed
### Description

When running the migrations, the user passwords are not hashed in the table. This means the users cannot authenticate because when checking for a correct login, the entered password is hashed and compared with what is in the table.

### Steps to reproduce the Bug

1. Run the migrations
2. Look in the db in the users table and see the passwords in plaintext

### Comment

_No response_

### PR DESCRIPTION
**Pull request by suiluj482**
Fixes #32 and fixes #89.
- API endpoints for login, logout, register, session
- added https://github.com/Barbapapazes/nuxt-authorization

### CODE REVIEW

File: server/api/auth/login.post.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        { email: c.usernameOrEmail },
+      ],
+    },
+  })
+
+  if (user == null) {
+    authLogger.debug("User not found", { usernameOrEmail: c.usernameOrEmail })
+    throw error
+  }
+  const match = bcrypt.compareSync(c.password, user.passwordHash)
```
Comment (niklhut) [inline server/api/auth/login.post.ts:44]: Use await here to avoid blocking

File: server/api/auth/register.post.ts
Comment (niklhut) [inline server/api/auth/register.post.ts:1]: I don't want two user create remove the other one or combine them

File: server/database/migrations/00_session.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      primaryKey: true,
+      defaultValue: DataTypes.UUIDV4,
+    },
+    // subject
+    sub: {
+      type: DataTypes.UUID,
+      // references: {
+      //   model: User,
+      //   key: "id",
+      // },
```
Comment (niklhut) [inline server/database/migrations/00_session.js:19]: Remove code comment

File: server/database/models/Session.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { DataTypes, Model } from "sequelize"
+import User from "./User"
+
+class Session extends Model {
+  declare id: string
+  declare sub: string
+  // declare exp: Date
```
Comment (niklhut) [inline server/database/models/Session.ts:7]: Remove code comment

File: shared/utils/abilities.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ */
+export const testAbillity = defineAbility({ allowGuest: true }, _ => true)
+
+// export const listPosts = defineAbility(() => true) // Only authenticated users can list posts
+
+// export const editPost = defineAbility((user: User, post: Post) => {
+//   return user.id === post.authorId
+// })
+
+// export const listPosts2 = defineAbility({ allowGuest: true }, (user: User | null) => true)
```
Comment (niklhut) [inline shared/utils/abilities.ts:16]: Remove code comment

**Review by niklhut — CHANGES_REQUESTED** [review]

File: server/database/models/SessionToken.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    },
+    // session: {
+    //   type: DataTypes.UUID,
+    //   references: {
+    //     model: Session,
+    //     key: "id",
+    //   },
+    //   allowNull: false,
+    // },
+    valid: {
```
Comment (niklhut) [inline server/database/models/SessionToken.ts:32]: Code comment

File: shared/utils/abilities.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * Ability to edit experiment
+ */
+export const canEditExperiment = defineAbility((user: UserDetail, experiment: ExperimentList) => {
+  return user.role === "Administrator" || user.id === experiment.userId
+})
+/**
+ * Ability to see experiment
+ */
+export const canSeeExperiment = defineAbility({ allowGuest: true }, (user: UserDetail, experiment: ExperimentList) => {
+  return experiment.status === "Accepted" || (user.role === "Moderator" || user.id === experiment.userId)
```
Comment (niklhut) [inline shared/utils/abilities.ts:22]: An admin could not see an non-accepted experiment

File: server/database/models/Session.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    // subject
+    // sub: {
+    //   type: DataTypes.UUID,
+    //   references: {
+    //     model: User,
+    //     key: "id",
+    //   },
+    //   allowNull: false,
+    // },
+    // expiration time
```
Comment (niklhut) [inline server/database/models/Session.ts:30]: Code comment

File: server/api/auth/login.post.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+const error = createError({
+  status: 401,
+  message: "Username, email or password wrong",
+})
+
+export default defineEventHandler(async (event) => {
+  const c = await readValidatedBody(event, body => v.parse(schema, body))
+
+  const whereClause
+    = v.is(v.pipe(v.string(), v.email()), c.usernameOrEmail)
```
Comment (niklhut) [inline server/api/auth/login.post.ts:22]: Maybe just use a (local) helper function isEmail then this is more readable and the ternary operator is smaller and teamscale should accept it like it did in legal

File: server/utils/user.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    v.regex(/[A-Z]/, "Password must contain at least one uppercase letter"),
+    v.regex(/[0-9]/, "Password must contain at least one number"),
+  ),
+}
+
+/**
+ * Finds a user by either their username or id.
+ *
+ * @param {string} usernameOrId The username or id of the user to search for.
+ * @returns {Promise<User>} A promise which resolves to the user object if found, or rejects with a 404 error if not found.
```
Comment (niklhut) [inline server/utils/user.ts:53]: Just add a linebreak somewhere

**Review by niklhut — CHANGES_REQUESTED** [review]
Looks good but a few minor changes. Also the tests will need to be extended to cover the negative cases in a different pr

**Review by niklhut — APPROVED** [review]
LGTM
